### PR TITLE
Add a TransformFunctionDefinition dataclass

### LINKED ISSUE #7610: Convert transfer_function_definitions in gen_kw_config from list of strings
`transfer_function_definitions´ is a list of strings that need to be parsed each time, we should instead make this a dataclass/pydantic baseclass so we dont have to parse a string:

Image

Will require migration.

### PR DESCRIPTION
**Pull request by frode-aarstad**
**Issue**
Resolves #7610

**Approach**
_Short description of the approach_

(Screenshot of new behavior in GUI if applicable)

- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Added appropriate release note label
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Make sure tests pass locally (after every commit!)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [ ] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Create Backport PR to latest release

### CODE REVIEW

File: src/ert/config/gen_kw_config.py [hunk 1]
```diff
@@ -58,18 +59,44 @@ def _get_abs_path(file: Optional[str]) -> Optional[str]:
     return file
 
 
+@dataclass
+class TransformFunctionDefinition:
+    func_name: str
```
Comment (dafeda) [inline src/ert/config/gen_kw_config.py:64]: Can we drop the `func` from `fun_name` and `param_func_name`?

File: src/ert/config/gen_kw_config.py [hunk 2]
```diff
@@ -58,18 +59,44 @@ def _get_abs_path(file: Optional[str]) -> Optional[str]:
...
-    transfer_function_definitions: List[str]
+    transfer_function_definitions: List[TransformFunctionDefinition]
     forward_init_file: Optional[str] = None
 
     def __post_init__(self) -> None:
         self.transfer_functions: List[TransferFunction] = []
         for e in self.transfer_function_definitions:
-            self.transfer_functions.append(self._parse_transfer_function(e))
-
+            if isinstance(e, str):
```
Comment (dafeda) [inline src/ert/config/gen_kw_config.py:79]: Given that we have to check whether `transfer_function_definitions` is a `str` or `dict`, does it make sense to type it as `List[TransformFunctionDefinition]`?

File: src/ert/config/gen_kw_config.py [hunk 3]
```diff
@@ -404,64 +438,44 @@ def _sample_value(
...
-                parameter_list=params,
-                calc_func=PRIOR_FUNCTIONS[param_func_name],
-            )
+        params = dict(zip(param_names, param_floats))
 
-        else:
-            raise ConfigValidationError(
-                f"Too few instructions provided in: {param_string}"
-            )
+        return TransferFunction(
```
Comment (dafeda) [inline src/ert/config/gen_kw_config.py:473]: Perhaps rename to `TransformFunction`. Whenever I see `TransferFunction` I think of signal processing...

File: src/ert/config/gen_kw_config.py [hunk 4]
```diff
@@ -404,64 +438,44 @@ def _sample_value(
...
+    ) -> TransferFunction:
+        if t.param_func_name is None and t.values is None:
+            raise ConfigValidationError(-f"Too few instructions provided in: {t}")
+
+        if (
+            t.param_func_name not in TRANS_FUNC_ARGS
+            or t.param_func_name not in PRIOR_FUNCTIONS
+        ):
+            raise ConfigValidationError(
+                f"Unknown transfer function provided: {t.param_func_name}"
```
Comment (dafeda) [inline src/ert/config/gen_kw_config.py:452]:
```suggestion
                f"Unknown transform function provided: {t.param_func_name}"
```

File: src/ert/config/gen_kw_config.py [hunk 5]
```diff
@@ -611,3 +625,18 @@ def calculate(self, x: float, arg: List[float]) -> float:
     "CONST": TransferFunction.trans_const,
     "RAW": TransferFunction.trans_raw,
 }
+
+
+TRANS_FUNC_ARGS: dict[str, List[str]] = {
```
Comment (dafeda) [inline src/ert/config/gen_kw_config.py:630]: Perhaps we should consider not using the term ARGS (arguments).
To me, this is a dictionary which maps distribution types to a list of their respective `parameters`.
Is `DISTRIBUTION_PARAMETERS` perhaps a better name?

File: src/ert/config/gen_kw_config.py (hunk 1 above)
Comment (frode-aarstad) [reply src/ert/config/gen_kw_config.py:64]: I will drop the funk

File: src/ert/config/gen_kw_config.py (hunk 2 above)
Comment (frode-aarstad) [reply src/ert/config/gen_kw_config.py:79]: I will add a Union

File: tests/unit_tests/storage/test_parameter_sample_types.py [hunk 6]
```diff
@@ -221,15 +222,24 @@ def test_that_first_three_parameters_sampled_snapshot(tmpdir, storage):
     [
         (
             "MY_KEYWORD <MY_KEYWORD>\nMY_SECOND_KEYWORD <MY_SECOND_KEYWORD>",
-            ["MY_KEYWORD NORMAL 0 1", "MY_SECOND_KEYWORD NORMAL 0 1"],
+            # ["MY_KEYWORD NORMAL 0 1", "MY_SECOND_KEYWORD NORMAL 0 1"],
```
Comment (dafeda) [inline tests/unit_tests/storage/test_parameter_sample_types.py:225]:
```suggestion
```

File: tests/unit_tests/storage/test_parameter_sample_types.py [hunk 7]
```diff
@@ -221,15 +222,24 @@ def test_that_first_three_parameters_sampled_snapshot(tmpdir, storage):
...
+            [
+                TransformFunctionDefinition("MY_KEYWORD", "NORMAL", [0, 1]),
+                TransformFunctionDefinition("MY_SECOND_KEYWORD", "NORMAL", [0, 1]),
+            ],
         ),
         (
             "MY_KEYWORD <MY_KEYWORD>",
-            ["MY_KEYWORD NORMAL 0 1"],
+            # ["MY_KEYWORD NORMAL 0 1"],
```
Comment (dafeda) [inline tests/unit_tests/storage/test_parameter_sample_types.py:233]:
```suggestion
```

File: tests/unit_tests/storage/test_parameter_sample_types.py [hunk 8]
```diff
@@ -221,15 +222,24 @@ def test_that_first_three_parameters_sampled_snapshot(tmpdir, storage):
...
+            [TransformFunctionDefinition("MY_KEYWORD", "NORMAL", [0, 1])],
         ),
         (
             "MY_FIRST_KEYWORD <MY_FIRST_KEYWORD>\nMY_KEYWORD <MY_KEYWORD>",
-            ["MY_FIRST_KEYWORD NORMAL 0 1", "MY_KEYWORD NORMAL 0 1"],
+            # ["MY_FIRST_KEYWORD NORMAL 0 1", "MY_KEYWORD NORMAL 0 1"],
```
Comment (dafeda) [inline tests/unit_tests/storage/test_parameter_sample_types.py:238]:
```suggestion
```

**Pull request merged**